Refresh plugin for September 2023

basil opened the pull request.

basil commented on .mvn/extensions.xml:
```diff
@@ -0,0 +1,7 @@
+<extensions xmlns="http://maven.apache.org/EXTENSIONS/1.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/EXTENSIONS/1.0.0 http://maven.apache.org/xsd/core-extensions-1.0.0.xsd">
```
As in the archetype.

basil commented on .mvn/maven.config:
```diff
@@ -0,0 +1,2 @@
+-Pconsume-incrementals
```
As in the archetype.

basil commented on Jenkinsfile:
```diff
@@ -1,6 +1,6 @@
   configurations: [
-    [platform: 'linux', jdk: 8],
-    [platform: 'windows', jdk: 8],
+    [platform: 'linux', jdk: 21],
```
As in the archetype.

basil commented on pom.xml:
```diff
@@ -5,16 +5,16 @@
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>4.37</version>
+    <version>4.73</version>
```
Latest version at the time of this writing.

basil commented on pom.xml:
```diff
@@ -5,16 +5,16 @@
   <groupId>com.amadeus.jenkins.plugins</groupId>
   <artifactId>workflow-cps-global-lib-http</artifactId>
-  <version>2.32.0-SNAPSHOT</version>
+  <version>${revision}${changelist}</version>
```
As in the archetype.

basil commented on pom.xml:
```diff
@@ -5,16 +5,16 @@
   <packaging>hpi</packaging>
   <name>Pipeline: Shared Groovy Libraries through HTTP retrieval</name>
-  <url>https://github.com/jenkinsci/workflow-cps-global-lib-http-plugin</url>
+  <url>https://github.com/jenkinsci/${project.artifactId}-plugin</url>
```
As in the archetype.

basil commented on pom.xml:
```diff
@@ -37,34 +37,28 @@
   </pluginRepositories>
 
   <properties>
+    <revision>2.32.0</revision>
+    <changelist>-SNAPSHOT</changelist>
```
As in the archetype.

basil commented on pom.xml:
```diff
@@ -37,34 +37,28 @@
-    <java.level>8</java.level>
-    <artifactory-maven-plugin.version>2.6.1</artifactory-maven-plugin.version>
-    <project.build.testSourceDirectory>src/test/java</project.build.testSourceDirectory>
+    <jenkins.version>2.387.3</jenkins.version>
```
https://www.jenkins.io/doc/developer/plugin-development/choosing-jenkins-baseline/

fabricepipart1a replied:
Indeed, I was worried that you took a too recent baseline. It is actually relatively reasonable and still covers most of the user base of the plugin as per https://stats.jenkins.io/pluginversions/workflow-cps-global-lib-http.html

basil commented on pom.xml:
```diff
@@ -37,34 +37,28 @@
         <groupId>io.jenkins.tools.bom</groupId>
-        <artifactId>bom-2.289.x</artifactId>
-        <version>1181.v04b_21d4b_0d6c</version>
+        <artifactId>bom-2.387.x</artifactId>
+        <version>2423.vce598171d115</version>
```
Latest at the time of this writing.

basil commented on pom.xml:
```diff
@@ -37,34 +37,28 @@
     <dependency>
-      <groupId>org.jenkins-ci.plugins.workflow</groupId>
-      <artifactId>workflow-cps-global-lib</artifactId>
+      <groupId>io.jenkins.plugins</groupId>
+      <artifactId>pipeline-groovy-lib</artifactId>
```
This plugin was split into two plugins, one deprecated and one non-deprecated. Switching this plugin's dependency to the non-deprecated one.

basil commented on pom.xml:
```diff
@@ -89,34 +83,14 @@
       <version>3.24.2</version>
       <scope>test</scope>
     </dependency>
-    <dependency>
-      <groupId>org.hamcrest</groupId>
-      <artifactId>hamcrest-core</artifactId>
-      <version>2.2</version>
-      <scope>test</scope>
-    </dependency>
-    <dependency>
-      <groupId>org.hamcrest</groupId>
-      <artifactId>hamcrest-library</artifactId>
-      <version>2.2</version>
-      <scope>test</scope>
-    </dependency>
```
Redundant; this is in the plugin parent POM.

basil commented on pom.xml:
```diff
@@ -89,34 +83,14 @@
-      <groupId>junit</groupId>
-      <artifactId>junit</artifactId>
-      <version>4.13.2</version>
+      <artifactId>mockito-core</artifactId>
```
New artifact ID upstream

basil commented on pom.xml:
```diff
@@ -89,34 +83,14 @@
-      <scope>test</scope>
-    </dependency>
-    <dependency>
-      <groupId>junit</groupId>
-      <artifactId>junit</artifactId>
-      <version>4.13.2</version>
```
Redundant; specified already in the parent PMO

basil commented on pom.xml:
```diff
@@ -89,34 +83,14 @@
     <dependency>
       <groupId>io.jenkins</groupId>
       <artifactId>configuration-as-code</artifactId>
-      <version>1.55.1</version>
```
Managed in the BOM these days

basil commented on pom.xml:
```diff
@@ -136,51 +110,15 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins.workflow</groupId>
       <artifactId>workflow-job</artifactId>
-      <version>2.32</version>
```
Managed in the BOM these days

basil commented on pom.xml:
```diff
@@ -136,51 +110,15 @@
     </dependency>
   </dependencies>
 
-  <build>
-    <plugins>
-      <plugin>
-        <groupId>org.jacoco</groupId>
-        <artifactId>jacoco-maven-plugin</artifactId>
-        <executions>
-          <execution>
-            <id>default-prepare-agent</id>
-            <goals>
-              <goal>prepare-agent</goal>
-            </goals>
-          </execution>
-          <execution>
-            <id>default-report</id>
-            <phase>prepare-package</phase>
-            <goals>
-              <goal>report</goal>
-            </goals>
-          </execution>
-          <execution>
-            <id>default-check</id>
-            <goals>
-              <goal>check</goal>
-            </goals>
-            <configuration>
-              <rules>
-              </rules>
-            </configuration>
-          </execution>
-        </executions>
-      </plugin>
-
-    </plugins>
-  </build>
```
Redundant; already in the plugin parent POM in the `enable-jacoco` profile (enabled in CI).

basil commented on pom.xml:
```diff
@@ -136,51 +110,15 @@
-    <developerConnection>scm:git:${project.scm.url}</developerConnection>
-    <url>https://github.com/jenkinsci/workflow-cps-global-lib-http-plugin.git</url>
-    <tag>2.18.0</tag>
+    <connection>scm:git:https://github.com/${gitHubRepo}.git</connection>
+    <developerConnection>scm:git:git@github.com:${gitHubRepo}.git</developerConnection>
+    <url>https://github.com/${gitHubRepo}</url>
+    <tag>${scmTag}</tag>
```
As in the archetype.

basil commented on src/test/java/com/amadeus/jenkins/plugins/workflow/libs/ConfigurationRoundtripTest.java:
```diff
@@ -88,14 +88,13 @@ public void testFormatDidNotChange() throws Exception {
-        }
-        return previousConfig;
+        String config = Files.readString(configFile.toPath(), StandardCharsets.UTF_8);
+        return config.replaceAll(
+                "<org.jenkinsci.plugins.workflow.libs.GlobalLibraries plugin=\"pipeline-groovy-lib@.*\">",
+                "<org.jenkinsci.plugins.workflow.libs.GlobalLibraries>");
```
Recent versions of core embed the plugin version in the XML. Since this would be changing frequently as Dependabot updates the plugin BOM, strip the version to ensure test stability.

basil commented on src/test/java/com/amadeus/jenkins/plugins/workflow/libs/HttpRetrieverIntegrationTests.java:
```diff
@@ -225,14 +227,18 @@ FilePath buildJobWithLibrary(@Nullable String fixtureName, @Nullable String libr
 
         WorkflowRun run = j.buildAndAssertSuccess(p);
-        return new FilePath(run.getRootDir().toPath().resolve("libs").resolve(libraryName).toFile());
+        List<Path> children = Files.list(run.getRootDir().toPath().resolve("libs"))
+                .filter(Files::isDirectory)
+                .collect(Collectors.toList());
+        assertThat(children, hasSize(1));
+        return children.get(0);
```
Newer versions of `pipeline-groovy-lib` name this directory with a hash rather than the library name. Since it's the only child that is a directory, it's still easy to fish it out.

basil commented on src/test/java/com/amadeus/jenkins/plugins/workflow/libs/HttpRetrieverTest.java:
```diff
@@ -24,7 +24,7 @@
 import org.mockito.Mock;
 import org.mockito.Mockito;
-import org.mockito.runners.MockitoJUnitRunner;
+import org.mockito.junit.MockitoJUnitRunner;
```
Package rename upstream.

jglick commented on src/test/java/com/amadeus/jenkins/plugins/workflow/libs/HttpRetrieverIntegrationTests.java:
```diff
@@ -225,18 +230,23 @@ FilePath buildJobWithLibrary(@Nullable String fixtureName, @Nullable String libr
 
     @Test
     public void retrievesWithPreemptiveAuthIfNeeded() throws Exception {
+        Assume.assumeFalse("TODO flakes on Windows due to open file handles", Functions.isWindows());
```
```suggestion
        Assume.assumeFalse("TODO fails on Windows", Functions.isWindows());
```
https://github.com/jenkinsci/jenkins-test-harness/pull/654

jglick commented on src/test/java/com/amadeus/jenkins/plugins/workflow/libs/HttpRetrieverIntegrationTests.java:
```diff
@@ -163,40 +171,37 @@ public void exceptionWhenRetrieveIncorrectUrl() throws Exception {
-
-        String libraryName = "foo";
-
+        Assume.assumeFalse("TODO flakes on Windows due to open file handles", Functions.isWindows());
```
```suggestion
        Assume.assumeFalse("TODO fails on Windows", Functions.isWindows());
```

basil commented:
Hi @jglick, thank you for taking the time to review this change. Although I appreciate the time you spent reviewing this change, I won't be addressing your feedback. My request for @fabricepipart1a to merge this PR remains.

fabricepipart1a commented:
Thanks for the PR. I'll review ASAP

fabricepipart1a reviewed (approved):
The contribution is immensely appreciated!
I see nothing that would prevent this to be released. It is work I would have eventually needed to do.

Thanks a lot for the contribution

jtnord commented on src/test/java/com/amadeus/jenkins/plugins/workflow/libs/HttpRetrieverIntegrationTests.java:
```diff
@@ -225,18 +230,23 @@ FilePath buildJobWithLibrary(@Nullable String fixtureName, @Nullable String libr
 
         WorkflowRun run = j.buildAndAssertSuccess(p);
-        return new FilePath(run.getRootDir().toPath().resolve("libs").resolve(libraryName).toFile());
+        List<Path> children = Files.list(run.getRootDir().toPath().resolve("libs"))
+                .filter(Files::isDirectory)
+                .collect(Collectors.toList());
```
this creates a stream holding a file handle - and it is never closed, this is likely a cause of the issue on windows where the clean-up after the test can not happen correctly.

Terminal operators do not close underlying streams.

jtnord commented on src/test/java/com/amadeus/jenkins/plugins/workflow/libs/HttpRetrieverIntegrationTests.java:
```diff
@@ -261,12 +271,16 @@ public void retrievesWithPreemptiveAuthIfNeeded() throws Exception {
-        Assert.assertTrue(target.child("src").exists());
-        Assert.assertTrue(target.child("vars").exists());
-        Assert.assertTrue(target.child("resources").exists());
+        List<Path> children = Files.list(run.getRootDir().toPath().resolve("libs"))
```
this creates a stream holding a file handle - and it is never closed, this is likely a cause of the issue on windows where the clean-up after the test can not happen correctly.

Terminal operators do not close underlying streams.

jtnord reviewed (commented):
streams need to be closed when they are operating on operating system resources.
